n2y/plugins/internallinks.py

```diff
@@ -0,0 +1,68 @@
+import typing
+import uuid
+import urllib.parse
+
+from n2y.blocks import Block, ChildPageBlock
+from n2y.logger import logger
+from n2y.page import Page
+from n2y.rich_text import TextRichText
+from n2y.utils import header_id_from_text
+
+
+def get_notion_id_from_href(href: str) -> typing.Optional[str]:
+    """Extract the ID of a target block from a href fragment"""
+    target_id = urllib.parse.urlparse(href, allow_fragments=True).fragment
+    if target_id is None:
+        return None
+    # Convert to UUID and back to ensure the format is correct
+    try:
+        return str(uuid.UUID(target_id))
+    except ValueError:
+        # fragment is not a Notion ID
+        return None
+
+
+def is_internal_link(href: typing.Optional[str], notion_id: str) -> bool:
+    """Is the href fragment a link to a block on the same page?"""
+    return href is not None and href.startswith(f"/{notion_id.replace('-', '')}")
+
+
+def find_target_block(page: Page, target_id: str) -> Block:
+    page_block: ChildPageBlock = page.block
+    if page_block is None:
+        raise ValueError("Page.block is None")
+
+    try:
+        return next(
+            child for child in page_block.children if child.notion_id == target_id
+        )
+    except StopIteration:
+        logger.error(f"Page missing block with target id '{target_id}'")
+        raise
+
+
+class NotionInternalLink(TextRichText):
+    """ This plugin fixes internal links to headers on the same page.
+
+    Instead of leaving the link to point to the original notion source page,
+    it substitutes the link to instead be the ID of the header.
+    """
+
+    def to_pandoc(self):
+        if not is_internal_link(self.href, self.block.page.notion_id):
+            return super().to_pandoc()
```

Comment:
Is there a reason why we couldn't check whether it's an internal link in the `__init__` and then raise a `UseNextClass` exception if it's not? This would allow people to have alternate plugins for the `TextRichText` class that could apply for non-internal links.

Reply:
That's a good idea. It breaks the tests though, but it's probably because of how the test constructs the test page and its contents. I'll have to look into it.

Reply:
@notmatthancock, about the concern you raised during our call, I would note that the block may be `None` whether `is_internal_link` is called in `to_pandoc` or `__init__.` To mitigate this risk, I'd write something like the following for `__init__`:
```py
    def __init__(self, client, notion_data, block):
        super().__init__(client, notion_data, block)
        if not block or not is_internal_link(self.href, self.block.page.notion_id):
            raise UseNextClass
```

Reply:
https://github.com/innolitics/n2y/pull/195/commits/c56fdd714d00b67d643a3cf5d16d4849b88b15c3